doc/dev/mds_internals/locking.rst

```diff
@@ -0,0 +1,143 @@
+
+.. note:: MDS also defines `remote_wrlock` which is primarly used during rename operation when the destination dentry is on another (active) MDS than the source MDS.
+
+Lock States and Lock State Machine
+----------------------------------
+
+MDS defines various lock states (defined in `src/mds/locks.h` source). Not all lock states are valid for a given lock class. Each lock class defines its own lock transition rules and are organized as Lock State Machines. The lock states (`LOCK_*`) are not locks themselves, but control if a lock is allowed to be taken. Each state follows `LOCK_<STATE>` or `LOCK_<FROM_STATE>_<TO_STATE>` naming terminology and can be summed up as::
+
+  LOCK_SYNC  - anybody can read lock, no one can write lock and exclusive lock
+  LOCK_LOCK  - no one can read lock, only primary mds can write lock or exclusive lock
+  LOCK_MIX   - anybody can write lock, no one can read lock or exclusive lock
+  LOCK_XLOCK - someone is holding a exclusive lock
+
+Other lock states (such as `LOCK_XSYN`, `LOCK_TSYN`, etc..) are additional states that are defined as an optimization for certain client behaviour (`LOCK_XSYN` allows clients to keep the buffered writes and not flush it to the OSDs and temporarily pausing writes).
+
+Intermediate lock states (`LOCK_<FROM_STATE>_<TO_STATE>`) denote transition of a lock from one state (`<FROM_STATE>`) to another (`<TO_STATE>`).
+
+Each lock class defines its own Lock State Machine and can be found in `src/mds/locks.c` source. The state machines are explained when discussing Lock Transition in the section below.
+
+Lock Transition
+---------------
+
+Transition of lock from one state to another is mostly prompted by a (client) request or a change that the MDS is undergoing, such as tree migration. Let's consider a simple case of two clients: One client does a `stat()` (`getattr()` or `lookup()`) to fetch UID/GID of a inode, and the other client does a `setattr()` to change the UID/GID of the same inode. The first client (most likely) has `As` (iauth shared) caps issued to it by the MDS. Now, when the other client does a `setattr()` call to the MDS, the MDS adds a `xlock` to the inodes' `authlock` (`CEPH_LOCK_IAUTH`)::
+
+  Server::handle_client_setattr()
+      if (mask & (CEPH_SETATTR_MODE|CEPH_SETATTR_UID|CEPH_SETATTR_GID|CEPH_SETATTR_BTIME|CEPH_SETATTR_KILL_SGUID))
+        lov.add_xlock(&cur->authlock);
+
+Note that the MDS adds a bunch of other locks for this inode, but for now let's only work on IAUTH. Now, `CEPH_LOCK_IAUTH` is a `SimpleLock` class, and its lock transition state machine is::
+
+                      // stable     loner  rep state  r     rp   rd   wr   fwr  l    x    caps,other
```

Comment:
My understanding is that the `r`, `rp`, `rd`, `wr`, `fwr`, `l` and `x` above are describing the relation between the MDSes. And the `caps` and `other` are describing the relation between clients.

And also both the above two type membters are also describing the relations between MDS and client.

Such as, if the `x` is AUTH, that means only the auth MDS could xlock it and the client related membters, which are `caps` and `other` must be zero. That's because when a MDS want to xlock a Inode, no client could read/write to it in remote clients.

If my understanding is correct, IMO it's also worth to add some explaination about this.

Reply:
> My understanding is that the `r`, `rp`, `rd`, `wr`, `fwr`, `l` and `x` above are describing the relation between the MDSes.

Right - It controls which lock operation an MDS is allowed to perform in a (lock) state.

> And the `caps` and `other` are describing the relation between clients.

They describe which caps are allowed on clients when the MDS is in a (lock) state. Locker::Issue_caps checks for allowed caps for a particular lock mode and sends cap revoke messages to clients.

> 
> And also both the above two type membters are also describing the relations between MDS and client.
> 
> Such as, if the `x` is AUTH, that means only the auth MDS could xlock it and the client related membters, which are `caps` and `other` must be zero. That's because when a MDS want to xlock a Inode, no client could read/write to it in remote clients.

Yeh - the zero (0) is (possibly) telling that caps must be revoked from the clients.

> 
> If my understanding is correct, IMO it's also worth to add some explanation about this.

This is explained (to some extent) in the paragraph below. Can be elaborated more....